url284ba87ea1_swabhs_forbidden_paths_tgz-pJ8-tagging_hypergraph_TaggingHypergraphGeneratorJ8/src/semiring/KBestSemiringSmart.java

```diff
@@ -0,0 +1,63 @@
+package semiring;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import utility.MaxPriorityQ;
+import utility.PositionVector;
+
+public class KBestSemiringSmart extends KBestSemiring {
+
+	private int k;
+	
+	public KBestSemiringSmart(int k) {
+		super(k);
+		this.k = k;
+	}
+	
+	/**
+	 * Given a list of lists, each of size at most k, merge them and return a single list of size 
+	 * at most k, containing the largest elements
+	 */
+	public List<Derivation> multiply(List<List<Derivation>> derivationsSet) {
+		List<Derivation> kbest = new ArrayList<Derivation>();
+		System.out.println(derivationsSet.size());
+		PositionVector pVector = new PositionVector(-1, derivationsSet.size());
+		kbest.add(SemiringUtils.getCandidateDerivation(derivationsSet, pVector, null));
+		
+		Map<Derivation, PositionVector> nextDerivations = new HashMap<Derivation, PositionVector>();
+		MaxPriorityQ q = new MaxPriorityQ();
+		
+		while (kbest.size() < k ) {
+			for (int i = 0; i < pVector.size(); i++) {
+				PositionVector candidatePosition = 
+						pVector.add(new PositionVector(i, pVector.size()));
+				Derivation candidateDerivation = 
+						SemiringUtils.getCandidateDerivation(derivationsSet, candidatePosition, null);
+				
+				if (candidateDerivation != null && !q.contains(candidateDerivation)) {
+					nextDerivations.put(candidateDerivation, candidatePosition);
+					q.insert(candidateDerivation);
+				}				
+			}
+			
+			if (q.size() == 0) {
+				break;
+			}
+			Derivation best = q.extractMax();
+			
+			for (Derivation d : nextDerivations.keySet()) {
+				if (d.getScore().equals(best.getScore())) {
+					pVector = nextDerivations.get(d);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.